### ISSUE TITLE
GraphML import issue with attr.name

### ISSUE THREAD

**Issue by sleroy** [jgrapht/jgrapht#370]
Hi,

I encountered an issue with the GraphML import/export.
Basically I export a GraphML file from my graph as a serialization and I read it again after.

`
org.jgrapht.ext.ImportException: Failed to parse GraphML
	at org.jgrapht.ext.GraphMLImporter.importGraph(GraphMLImporter.java:254)
	at org.jgrapht.ext.GraphImporter.importGraph(GraphImporter.java:68)	
`

The following generated information is the problem "Vertex name" in attr.name. When I am removing the space, it works properly.

Before
`<key id="vertex_label_key" for="node" attr.name="Vertex Label" attr.type="string"/>`

After

`<key id="vertex_label_key" for="node" attr.name="VertexLabel" attr.type="string"/>`

I provided a test in attachment.

Tell me if I can help you.

Sylvain

**Comment by sleroy** [comment]
Unit test

**Comment by jkinable** [comment]
For some reason, I can see your unit test in the e-mail sent by github, but it doesn't show up in the issue above. Here' s a copy of your unit test:

```
@Test
public void testLoading2() throws ExportException, ImportException {
final DefaultDirectedGraph graph = new DefaultDirectedGraph<>(DefaultEdge.class);
graph.addVertex("A");
graph.addVertex("B");
graph.addVertex("C");

	// Writing in GraphML
	final ComponentNameProvider<String> nameProvider= component -> component;
	final GraphMLExporter<String, DefaultEdge> graphMLExporter = new GraphMLExporter<>(new IntegerComponentNameProvider(), nameProvider, new IntegerComponentNameProvider(), null);
	final File file1 = new File("/tmp/graph.graphml");
	graphMLExporter.exportGraph(graph, file1);

	// Reading again
	final VertexProvider<String> vertexProvider = (label, attributes) -> attributes.get("vertex_label_key");
	final DefaultDirectedGraph loadedGraph = new DefaultDirectedGraph<>(DefaultEdge.class);
	final ClassBasedEdgeFactory<String, DefaultEdge> edgeFactory = new ClassBasedEdgeFactory(DefaultEdge.class);
	final EdgeProvider<String, DefaultEdge> edgeProvider = (from, to, label, attributes) -> edgeFactory
			.createEdge(from, to);
	final GraphMLImporter<String, DefaultEdge> graphMLImporter = new GraphMLImporter<>(vertexProvider,
			edgeProvider);
	graphMLImporter.importGraph(loadedGraph, file1);

}
```

Thanks for reporting, we will look into it.

**Comment by jkinable** [comment]
I was able to reproduce this issue. 
```
org.jgrapht.io.ImportException: Failed to parse GraphML
	at org.jgrapht.io.GraphMLImporter.importGraph(GraphMLImporter.java:254)
	at org.jgrapht.io.GraphImporter.importGraph(GraphImporter.java:68)
	at org.jgrapht.io.GraphMLImporterTest.testLoading2(GraphMLImporterTest.java:1061)
	...
Caused by: org.xml.sax.SAXParseException; lineNumber: 2; columnNumber: 84; cvc-datatype-valid.1.2.1: 'Vertex Label' is not a valid value for 'NMTOKEN'.
```

**Comment by d-michail** [comment]
The GraphML that you are producing is non-valid. The error is that "'Vertex Label' is not a valid value for 'NMTOKEN'".

The importer validates using the GraphML schema, see for example [graphml-attributes.xsd](http://graphml.graphdrawing.org/xmlns/1.1/graphml-attributes.xsd) where you can see that the attr.name is restricted with xs:NMTOKEN.

NMTokens do not allow whitespace inside a string. See [here](http://stackoverflow.com/questions/5808976/what-is-the-usefulness-of-nmtoken-and-nmtokens-types) for some details.

The exporter (on-purpose) does not do any validation. We let the user take responsibility, as some users might want to produce something custom.

**Comment by d-michail** [comment]
Thanks for reporting. 

Just for completeness, if you want to validate on export, the [xerces2](https://xerces.apache.org/xerces2-j/javadocs/xerces2/org/apache/xerces/util/XMLChar.html#isValidNmtoken(java.lang.String)) library contains a method for checking that a string is a valid NMTOKEN.

**Comment by bpomykala** [comment]
I come up against the same problem. As I see this issue can be resolved by modifying some special attribute definitions in org.jgrapht.io.GraphMLExporter.java file.   [Link to file](https://github.com/jgrapht/jgrapht/blob/master/jgrapht-io/src/main/java/org/jgrapht/io/GraphMLExporter.java)  .

Current code:

`    // special attributes`
`    private static final String VERTEX_LABEL_DEFAULT_ATTRIBUTE_NAME = "Vertex Label";`
`    private static final String EDGE_WEIGHT_DEFAULT_ATTRIBUTE_NAME = "weight";`
`    private static final String EDGE_LABEL_DEFAULT_ATTRIBUTE_NAME = "Edge Label";`

Suggested fix:


`    // special attributes`
`    private static final String VERTEX_LABEL_DEFAULT_ATTRIBUTE_NAME = "VertexLabel";`
`    private static final String EDGE_WEIGHT_DEFAULT_ATTRIBUTE_NAME = "weight";`
`    private static final String EDGE_LABEL_DEFAULT_ATTRIBUTE_NAME = "EdgeLabel";`

**Comment by d-michail** [comment]
@bpomykala Thanks! Somehow I missed the part that our default attribute name was not valid!!! @sleroy Thanks for the original report.

Fixed it on PR #436 which includes other I/O package enhancements, to avoid merge conflicts.

**Comment by d-michail** [comment]
Closing this one after merging #436.
